### PR TITLE
Add "weight" parameter

### PR DESCRIPTION
**Pull request by pldin601**
Add `weight` parameter to adjust throttling depending on action "heaviness".

### CODE REVIEW

**Comment by JMPerez** [comment]
Hi @pldin601! Thanks for your PR.

I understand the use case but I think it can be covered today by creating 2 instances of the throttle object setting different requestPerSecond values. Or am I missing something?

**Comment by pldin601** [comment]
No, because is case of two instances we have two different throttles with own throttling queues. There are many APIs that requires to control request rate with specifying call weight, and it is bound to client IP or session.

File: lib/main.js [hunk 1]
```diff
@@ -18,15 +18,17 @@ function PromiseThrottle(options) {
 /**
  * Adds a promise
  * @param {Function} promise A function returning the promise to be added
+ * @param {number} weigth A "weight" of operation resolving by this promise
```
Comment (JMPerez) [inline lib/main.js]: Typo in `weigth` (should be `weight`)

File: lib/main.js [hunk 2]
```diff
@@ -18,15 +18,17 @@ function PromiseThrottle(options) {
...
  * @return {Promise} A promise
  */
-PromiseThrottle.prototype.add = function(promise) {
+PromiseThrottle.prototype.add = function(promise, weigth) {
```
Comment (JMPerez) [inline lib/main.js]: Same here

File: lib/main.js [hunk 3]
```diff
@@ -18,15 +18,17 @@ function PromiseThrottle(options) {
...
   var self = this;
   return new self.promiseImplementation(function(resolve, reject) {
     self.queued.push({
       resolve: resolve,
       reject: reject,
-      promise: promise
+      promise: promise,
+      weight: weigth || 1,
```
Comment (JMPerez) [inline lib/main.js]: And here (right one)

File: lib/main.js [hunk 4]
```diff
@@ -36,11 +38,12 @@ PromiseThrottle.prototype.add = function(promise) {
 /**
  * Adds all the promises passed as parameters
  * @param {Function} promises An array of functions that return a promise
+ * @param {number} weigth A "weight" of each operation resolving by array of promises
```
Comment (JMPerez) [inline lib/main.js]: Same here

File: lib/main.js [hunk 5]
```diff
@@ -36,11 +38,12 @@ PromiseThrottle.prototype.add = function(promise) {
...
  * @return {void}
  */
-PromiseThrottle.prototype.addAll = function(promises) {
+PromiseThrottle.prototype.addAll = function(promises, weigth) {
```
Comment (JMPerez) [inline lib/main.js]: Same here

File: lib/main.js [hunk 6]
```diff
@@ -36,11 +38,12 @@ PromiseThrottle.prototype.add = function(promise) {
...
   var addedPromises = promises.map(function(promise) {
-    return this.add(promise);
+    return this.add(promise, weigth || 1);
```
Comment (JMPerez) [inline lib/main.js]: Same here

File: lib/main.js [hunk 7]
```diff
@@ -53,7 +56,8 @@ PromiseThrottle.prototype.addAll = function(promises) {
 PromiseThrottle.prototype.dequeue = function() {
   if (this.queued.length > 0) {
     var now = new Date(),
-      inc = 1000 / this.requestsPerSecond,
+      weigth = this.queued[0].weight,
+      inc = (1000 / this.requestsPerSecond) * weigth,
```
Comment (JMPerez) [inline lib/main.js]: Same here

**Pull request merged**